lib/hmmlearn/base.py

```diff
@@ -9,13 +9,75 @@
 from sklearn.utils import check_array, check_random_state
 
 from . import _hmmc, _utils
+from .kl_divergence import kl_dirichlet
+from .stats import log_multivariate_normal_density
 from .utils import normalize, log_normalize, log_mask_zero
 
 
 _log = logging.getLogger(__name__)
 #: Supported decoder algorithms.
 DECODER_ALGORITHMS = frozenset(("viterbi", "map"))
 
+# Wrappers around _hmmc, used by both EM and VI
+def do_forward_scaling_pass(frameprob, startprob, transmat):
+    fwdlattice, scaling_factors = _hmmc.forward_scaling(
+        np.asarray(startprob), np.asarray(transmat),
```

Comment:
I just realized that pybind11 is perfectly happy to take array-likes and handle the casting itself (it ultimately uses PyArray_FromAny), which means that we don't need these np.asarray calls; I think it may(?) be easier to just get rid of these extra wrappers and e.g. adjust _hmmc.forward_scaling to return log_prob, fwdlattice, scaling_factors, or move log_mask_zero to the C++ side.  I can do that as a separate PR, if you prefer.

Reply:
Let me look at this - it will also simplify this MR.

Reply:
Done.